Source/Concurrency/InductiveSequentialization.cs

```diff
@@ -435,6 +440,293 @@ private List<Declaration> GenerateStepChecker(Action pendingAsync)
+          .ToList();
+
+      List<LinearityCheck> linearityChecks = new List<LinearityCheck>();
+      foreach (var domain in ltc.LinearDomains)
+      {
+       var existingExpr = Expr.True;
+       var inVars = new List<Variable>().Union(act.ModifiedGlobalVars)
+        .Where(x => LinearTypeChecker.InKinds.Contains(LinearTypeChecker.FindLinearKind(x))).Select(Expr.Ident).ToList();
+       var outVars = new List<Variable>().Union(act.ModifiedGlobalVars)
+        .Where(x => LinearTypeChecker.InKinds.Contains(LinearTypeChecker.FindLinearKind(x))).Select(Expr.Ident).ToList();
+       var inPermissionSet = ExprHelper.Old(ltc.UnionExprForPermissions(domain, ltc.PermissionExprs(domain, inVars)));
+       var outPermissionSet = ltc.UnionExprForPermissions(domain, ltc.PermissionExprs(domain, outVars));
+       var outSubsetInExpr = ltc.SubsetExprForPermissions(domain, inPermissionSet, outPermissionSet);
+       linearityChecks.Add(new LinearityCheck(
+          domain,
+          existingExpr,
+          outSubsetInExpr,
+          $"Only put permissions of type {domain.permissionType}",
+          $"only_put_perm_{act.Name}"));
+      }   
+      
+      List<Block> checkerBlocks = new List<Block>(linearityChecks.Count);
+      foreach (var lc in linearityChecks)
+      {
+        List<Cmd> cmds = new List<Cmd>(2);
+        if (lc.assume != null)
+        {
+          cmds.Add(CmdHelper.AssumeCmd(lc.assume));
+        }
+        cmds.Add(CmdHelper.AssertCmd(act.tok, lc.assert, lc.message));
+        var block = BlockHelper.Block($"{lc.domain.permissionType}_{lc.checkName}", cmds);
+        CivlUtil.ResolveAndTypecheck(civlTypeChecker.Options, block, ResolutionContext.State.Two);
+        checkerBlocks.Add(block);
+      }
+      
+      // Create init blocks
+      List<Block> blocks = new List<Block>(linearityChecks.Count + 1);
+      if(checkerBlocks.Count != 0){
+      blocks.Add(
+        BlockHelper.Block(
+          "init",
+          new List<Cmd>() { CmdHelper.CallCmd(act.Impl.Proc, inputs, outputs) },
+          checkerBlocks));
+      }
+      else{
+        blocks.Add(
+        BlockHelper.Block(
+          "init",
+          new List<Cmd>() { CmdHelper.CallCmd(act.Impl.Proc, inputs, outputs) }));
+      }
+      blocks.AddRange(checkerBlocks);
+
+      // Create the whole check procedure
+      string checkerName = civlTypeChecker.AddNamePrefix($"{rule}_SideCondition_{act.Name}");
+      Procedure linCheckerProc = DeclHelper.Procedure(checkerName,
+        inputs, outputs, requires, act.Impl.Proc.Modifies, new List<Ensures>());
+      Implementation linCheckImpl = DeclHelper.Implementation(linCheckerProc,
+        inputs, outputs,  new List<Variable>(), blocks);
+      return new List<Declaration>(new Declaration[] { linCheckerProc, linCheckImpl });
+    }
+
+    private List<Declaration> GenerateInconsistencyChecker(Action E1, Action E2)
+    {
+      var gateE1 = (Expr)Expr.True;
+      foreach(var g in E1.Gate)
```

Comment:
Replace loop with the version of Expr.And that takes in an Enumerable.

Same with the loops below.